File: src/net/eduard/permissions/manager/PermissionsPlayer.java
package net.eduard.permissions.manager;

import java.util.ArrayList;
import java.util.List;

import net.eduard.api.lib.game.FakePlayer;
import net.eduard.api.lib.storage.Storable;

@Storable.StorageAttributes(indentificate = true)
public class PermissionsPlayer {

	private FakePlayer player;
	private transient PermissionsPlayerEditor editor;
	private String prefix;
	private String suffix;

	private List<String> permissions = new ArrayList<>();
	@Storable.StorageAttributes(reference = true)
	private List<PermissionsGroup> groups = new ArrayList<>();

	public List<PermissionsGroup> getGroups() {
		return groups;
	}

	public void setGroups(List<PermissionsGroup> groups) {
		this.groups = groups;
	}

	public List<String> getPermissions() {
		return permissions;
	}

	public void setPermissions(List<String> permissions) {
		this.permissions = permissions;
	}

	public FakePlayer getPlayer() {
		return player;
	}

	public void setPlayer(FakePlayer player) {
		this.player = player;
	}

	public String getSuffix() {
		return suffix;
	}

	public void setSuffix(String suffix) {
		this.suffix = suffix;
	}

	public String getPrefix() {
		return prefix;
	}

	public void setPrefix(String prefix) {
		this.prefix = prefix;
	}

	public PermissionsPlayerEditor getEditor() {
		return editor;
	}

	public void setEditor(PermissionsPlayerEditor editor) {
		this.editor = editor;
	}

}

File: src/main/java/net/eduard/permissions/core/PermissionsPlayer.kt
package net.eduard.permissions.core

import net.eduard.api.lib.command.PlayerOffline
import net.eduard.api.lib.database.StorageManager
import net.eduard.api.lib.storage.Storable
import net.eduard.permissions.EduPermissions
import net.eduard.permissions.api.PermissionsAPI

@Storable.StorageAttributes(indentificate = true)
class PermissionsPlayer {

    @StorageManager.StoreKey
    var player = PlayerOffline("Eduard")

    @Transient
    lateinit var editor: PermissionsPlayerEditor
    var prefix = ""
    var suffix = ""
    var permissions = mutableListOf<String>()

    @Transient
    var groups = mutableListOf<PermissionsGroup>()
    var groupsNames = mutableListOf<String>()

    fun save() {
         EduPermissions.instance.storageManager
             .update(this)
    }


    fun hasPermission(permissionName: String): Boolean {
        for (perm in permissions) {
            if (perm == "*") {
                return true
            }
            if (perm.endsWith(".*")) {
                val permCortada = perm.replace(".*", "")
                if (permissionName.startsWith(permCortada)) {
                    return true
                }
            }
        }
        for (group in groups) {
            for (perm in group.permissions) {
                if (perm == "*") {
                    return true
                }
                if (perm.endsWith(".*")) {
                    val permCortada = perm.replace(".*", "")
                    if (permissionName.startsWith(permCortada)) {
                        return true
                    }
                }
            }
        }
        return false
    }

    fun canRemove(): Boolean {

        if (permissions.isEmpty()
            &&  groups.size == 1 && groupsNames.
            contains(PermissionsAPI.getInstance().groupDefault))
            return true


            return false
    }


}